aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -101,4 +104,199 @@ Tensor& randperm_out_cuda(Tensor& result, int64_t n, Generator* generator) {
   return result;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ triangle ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+namespace {
+// f: the number of elements in the first row of the trapezoid.
+// x: the index of the target coordinates ordered by row and then column.
+// Assume x corresponding to coordinates (row, col) in the trapezoid, where
+// row and col start from 0, then we have:
+//     (f + f + row - 1) * row / 2 <= x < (f + f + row) * (row + 1) / 2.    [*]
+// Therefore, row is the maximum integer satisfying the following inequality:
+//              (row + 2f - 1)row <= 2x
+//         row^2 + (2f-1)row - 2x <= 0.
+// Based on the formula of root, we have:
+// a = 1
+// b = 2f - 1
+// c = -2x
+// Use the root on the right, intuitively because it is where the left-hand side
+// formular flips from negative to positive. Full proof can be derived from the
+// second inequality in [*].
+// row = floor((-b + sqrt(b^2 - 4c)) / 2)
+// col = x - (f + f + row - 1) * row / 2
```

Comment:
This looks different from what is in the code below (you have `col = x - ((f + row - 1) * row >> 1)`)